mlflow/transformers.py

```diff
@@ -0,0 +1,909 @@
+            "Unable to store ModelCard data with the saved artifact. In order to "
+            "preserve this information, please install the huggingface_hub package "
+            "by running 'pip install huggingingface_hub>0.10.0'"
+        )
+
+
+def _build_pipeline_from_model_input(model, task: str):
+    """
+    Utility for generating a pipeline from component parts. If required components are not
+    specified, use the transformers library pipeline component validation to force raising an
+    exception. The underlying Exception thrown in transformers is verbose enough for diagnosis.
+    """
+    from transformers import pipeline
+
+    # Externalizing this configuration in case model has a task key.
+    pipeline_config = {
+        "task": task,
+        **model,
+    }
+    try:
+        return pipeline(**pipeline_config)
+    except Exception as e:
+        raise MlflowException(
+            "The provided model configuration cannot be created as a Pipeline. "
+            "Please verify that all required and compatible components are "
+            "specified with the correct keys.",
+            error_code=INVALID_PARAMETER_VALUE,
+        ) from e
+
+
+def _record_pipeline_components(pipeline) -> Dict[str, Any]:
+    """
+    Utility for recording which components are present in either the generated pipeline iff the
+    supplied save object is not a pipeline or the components of the supplied pipeline object.
+    """
+    components_conf = {}
+    components = []
+    if getattr(pipeline, "feature_extractor", None) is not None:
+        feature_extractor = pipeline.feature_extractor
+        components_conf.update({_FEATURE_EXTRACTOR_TYPE_KEY: _get_instance_type(feature_extractor)})
+        components.append(_FEATURE_EXTRACTOR_KEY)
+    if getattr(pipeline, "tokenizer", None) is not None:
+        tokenizer = pipeline.tokenizer
+        components_conf.update({_TOKENIZER_TYPE_KEY: _get_instance_type(tokenizer)})
+        components.append(_TOKENIZER_KEY)
+    if getattr(pipeline, "image_processor", None) is not None:
+        image_processor = pipeline.image_processor
+        components_conf.update({_IMAGE_PROCESSOR_TYPE_KEY: _get_instance_type(image_processor)})
+        components.append(_IMAGE_PROCESSOR_KEY)
+    if components:
+        components_conf.update({_COMPONENTS_BINARY_KEY: components})
+    return components_conf
+
+
+def _save_components(
+    root_path: pathlib.Path, component_config: Dict[str, Any], pipeline, processor
+):
+    """
+    Saves non-model pipeline components explicitly to a separate directory path for compatibility
+    with certain older model types. In earlier versions of ``transformers``, inferred feature
+    extractor types could be resolved to their correct processor types. This approach ensures
+    compatibility with the expected pipeline configuration argument assignments in later versions
+    of the ``Pipeline`` class.
+    """
+    component_types = component_config["components"]
```

Comment:
```suggestion
    component_types = component_config.get(_COMPONENTS_BINARY_KEY, [])
```

Can we use `_COMPONENTS_BINARY_KEY` here?

Reply:
yep! good catch